### PR TITLE
Update `reserved stock` class to handle concurrent requests

### LINKED ISSUE #1652: Checkout Block: Mitigate and test for concurrency issues when reserving stock
Introduced in #1425, the API creates draft orders during the checkout flow which also help reserve stock during checkout for a customer.

Core team have been working on a similar system for putting holds on stock which attempts to mitigate concurrency issues (too much stock being reserved if 2 or more requests come in simultaneously).

- https://github.com/woocommerce/woocommerce/pull/25172/files#diff-27a48ce67fa604181c90b4bb464164acR306
- https://github.com/woocommerce/woocommerce/pull/25172/

Since there is benefit to rolling out changes in blocks first, we should (with @woocommerce/proton and @vedanshujain) ensure the solution in blocks is flexible enough to be the solution used by core too, and ensure we have a testing tool to confirm this works.

TODO:
- [x] Review current stock reservation table solution, align with core
- [ ] Add concurrency testing tool
- [x] Implement protection against concurrency issues

### PR DESCRIPTION
**Pull request by mikejolley**
These changes solve the concurrency issue by borrowing certain code changes from https://github.com/woocommerce/woocommerce/pull/25172

The main changes are the atomic 'replace' to prevent stock being reserved for 2 orders at the same time, and the introduction of an `expires` column in the DB table.

Since this solution should ultimately be used in core too, I've created a core PR for the same class here https://github.com/woocommerce/woocommerce/pull/25708 Our version will be used during the transition to this (when we support wc 4.0+).

Fixes #1652 

### How to test the changes in this Pull Request:

Since the DB table changed, you'll need to do a little messing about in the DB. This hasn't shipped to users yet so no issues there.

1. Delete `wc_blocks_db_version` from your wp_options table.
2. Reactivate blocks.

This will ensure the `expires` column is added to your `wc_reserved_stock` table.

Then to test, follow the testing instructions in https://github.com/woocommerce/woocommerce/pull/25708, and on the Blocks side, create an order via the API 

- Add some items to your cart whilst logged in
- Basic auth to API for the same user.
- Do a post to /store/cart/order.
 - Check response.

The original issue has more details if needed https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1425

### CODE REVIEW

**Comment by mikejolley** [comment]
@nerrad If I put the legacy handling back (so it works even if core doesn't update), could we get this merged? The core PR has not been reviewed yet, but it's important we get this merged before we ship cart and checkout blocks. Not sure what to do.

**Comment by nerrad** [comment]
@mikejolley yes, let's put the legacy handling back so it works without a core merge. I'll review once that is in place.

**Comment by mikejolley** [comment]
Added the legacy function back, if the core class is not detected. If classnames change in core after the PR is merged this will need updating, but this should be ok to review now.

File: src/RestApi/StoreApi/Utilities/ReserveStockException.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @package WooCommerce/Blocks
+ */
+
+namespace Automattic\WooCommerce\Blocks\RestApi\StoreApi\Utilities;
+
+/**
+ * ReserveStockException class.
+ */
+class ReserveStockException extends \Exception {
```
Comment (nerrad) [inline src/RestApi/StoreApi/Utilities/ReserveStockException.php]: Big 👍 for custom exception.

File: tests/php/RestApi/StoreApi/Utilities/ReserveStock.php [hunk 2]
```diff
@@ -32,26 +32,46 @@ public function test_reserve_stock_for_order() {
...
-		$result = $class->reserve_stock_for_order( $order );
-		$this->assertTrue( $result );
+		$class->reserve_stock_for_order( $order );
 		$this->assertEquals( 8, $this->get_reserved_stock_by_product_id( $product->get_stock_managed_by_id() ) );
+	}
 
-		// Repeat again - should not be enough stock for this.
-		$order = OrderHelper::create_order( 1, $product );
+	/**
+	 * Test that stock is reserved for draft orders.
```
Comment (nerrad) [inline tests/php/RestApi/StoreApi/Utilities/ReserveStock.php]: Looks like copy pasta here. The description doesn't match the test.

**Pull request merged**